File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static final int INF = Integer.MAX_VALUE;

    static class Edge implements Comparable<Edge> {
        int to;
        int weight;

        Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge o) {
            return Integer.compare(this.weight, o.weight);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        int start = Integer.parseInt(br.readLine());

        List<List<Edge>> graph = new ArrayList<>();
        for (int i = 0; i <= v; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());
            graph.get(from).add(new Edge(to, weight));
        }

        int[] distance = new int[v + 1];
        Arrays.fill(distance, INF);
        distance[start] = 0;

        PriorityQueue<Edge> pq = new PriorityQueue<>();
        pq.offer(new Edge(start, 0));
        while (!pq.isEmpty()) {
            Edge cur = pq.poll();
            int curNode = cur.to;
            int curDist = cur.weight;

            if (curDist > distance[curNode]) continue;

            for (Edge edge : graph.get(curNode)) {
                int nextNode = edge.to;
                int newDist = distance[curNode] + edge.weight;

                if (newDist < distance[nextNode]) {
                    distance[nextNode] = newDist;
                    pq.offer(new Edge(nextNode, newDist));
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= v; i++) {
            if (distance[i] == INF) {
                sb.append("INF\n");
            } else {
                sb.append(distance[i]).append('\n');
            }
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/1753. 최단경로/최단경로.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.PriorityQueue

private val inf = Int.MAX_VALUE

data class Edge(val to: Int, val weight: Int) : Comparable<Edge> {
    override fun compareTo(other: Edge): Int = weight.compareTo(other.weight)
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (v, e) = br.readLine().split(" ").map { it.toInt() }
    val start = br.readLine().toInt()

    val graph = List(v + 1) { mutableListOf<Edge>() }
    repeat(e) {
        val (from, to, weight) = br.readLine().split(" ").map { it.toInt() }
        graph[from].add(Edge(to, weight))
    }

    val distance = IntArray(v + 1) { inf }
    distance[start] = 0

    val pq = PriorityQueue<Edge>()
    pq.offer(Edge(start, 0))
    while (pq.isNotEmpty()) {
        val (curNode, curDist) = pq.poll()

        if (curDist > distance[curNode]) continue

        for (edge in graph[curNode]) {
            val nextNode = edge.to
            val newDist = distance[curNode] + edge.weight

            if (newDist < distance[nextNode]) {
                distance[nextNode] = newDist
                pq.offer(Edge(nextNode, newDist))
            }
        }
    }

    val sb = StringBuilder()
    for (i in 1..v) {
        if (distance[i] == inf) {
            sb.append("INF\n")
        } else {
            sb.append(distance[i]).append('\n')
        }
    }

    print(sb)
}
